tests/cleanup_and_retention_test.go

```diff
@@ -0,0 +1,186 @@
+	}
+	start := time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)
+	startMilli := utils.UnixMilli(start)
+
+	scenarios := []struct {
+		limit        uint64
+		rowsToDelete uint64
+		rowsToSpare  uint64
+	}{
+		{10, 9, 100},
+		{10, 10, 100},
+		{10, 100, 100},
+		{10, 1000, 100},
+		{100, 100, 100},
+		{1000, 100000, 100},
+	}
+
+	for i := range scenarios {
+		scenario := scenarios[i]
+
+		t.Run(fmt.Sprintf("%d rows with limit %d", scenario.rowsToDelete, scenario.limit), func(t *testing.T) {
+			t.Parallel()
+
+			table := fmt.Sprintf("retention_test_limit_%d_delete_%d", scenario.limit, scenario.rowsToDelete)
+
+			db := database(t)
+
+			_, err := db.Exec(fmt.Sprintf(`CREATE TEMPORARY TABLE %s (event_time bigint(20) NOT NULL, PRIMARY KEY (event_time))`, table))
+			require.NoError(t, err)
+
+			// One row per millisecond.
+			total := scenario.rowsToDelete + scenario.rowsToSpare
+			values := make([]struct{ Time int64 }, 0, MaxPlaceholders)
+			for j := uint64(0); j < total; j++ {
+				values = append(values, struct{ Time int64 }{startMilli + int64(j)})
+
+				if len(values) == MaxPlaceholders-1 || j == total-1 {
+					_, err := db.NamedExec(fmt.Sprintf(`INSERT INTO %s (event_time) VALUES (:time)`, table), values)
+					require.NoError(t, err)
+					values = values[:0]
+				}
+			}
+
+			cleanup, err := db.CleanupOlderThan(
+				context.Background(),
+				icingadb.CleanupStmt{Table: table, Column: "event_time"},
+				scenario.limit,
+				start.Add(time.Millisecond*time.Duration(scenario.rowsToDelete)),
+			)
+			assert.NoError(t, err)
+
+			var count uint64
+			err = db.QueryRow(fmt.Sprintf(`SELECT COUNT(*) FROM %s`, table)).Scan(&count)
+			assert.NoError(t, err)
+			assert.Equal(t, int(scenario.rowsToSpare), int(count), "rows left in table")
+			assert.Equal(t, int(scenario.rowsToDelete), int(cleanup.Count), "deleted rows")
+			assert.Equal(t, int((scenario.rowsToDelete+scenario.limit-1)/scenario.limit), int(cleanup.Rounds), "number of rounds")
+		})
+	}
+
+	t.Run("History", func(t *testing.T) {
+		t.Parallel()
+
+		db := database(t)
+		retentionDays := 1
+		scenario := scenarios[len(scenarios)-1]
+		wg := sync.WaitGroup{}
+
+		for _, stmt := range history.RetentionStatements {
+			_, err := db.Exec(fmt.Sprintf(
+				`CREATE TABLE %s (%s int unsigned NOT NULL AUTO_INCREMENT, %s bigint unsigned NOT NULL, PRIMARY KEY (%s), INDEX idx_%s (%s))`,
+				stmt.Table, stmt.PK, stmt.Column, stmt.PK, stmt.Table, stmt.Column,
+			))
+			require.NoError(t, err)
+
+			total := scenario.rowsToDelete + scenario.rowsToSpare
+			values := make([]struct{ Time int64 }, 0, MaxPlaceholders)
+			start := time.Now().AddDate(0, 0, -retentionDays).Add(-1 * time.Millisecond * time.Duration(scenario.rowsToDelete))
+			startMilli := utils.UnixMilli(start)
+			for j := uint64(0); j < total; j++ {
+				if j == scenario.rowsToDelete {
+					startMilli += int64(time.Minute)
```

Comment:
The test would fail if inserting all test data would take longer than this, wouldn't it? In my one-off test, inserting took 15s, so there's some safety margin there, but would there be any downside in using a larger value here?